### DISCUSSION TITLE
Excluding fields from the GET /collection REST API results

### DISCUSSION THREAD

**Discussion by joncys** [payloadcms/payload#2794] [Q&A]
So, this seems like something that should've been answered or mentioned somewhere before, but I cannot find anything on the official docs or by searching and reading through the Q&A section of the forums here.

Is there a way to filter / exclude fields from the REST API response when. querying the collection, like when doing:

```
GET api/collection
```

I want only certain fields to be returned, like maybe imagined `title`, `publishedAt` and `coverImage`.
Imagine I also have a `content` field that is of type `richText`. I don't want that to be returned for when I just want to list the collection items and show their `coverImage` and `publishedAt` date.

I cannot really do `depth=0`, because I want to go in with `depth: 1` for the `coverImage` field and also, it does not remove most of the text for the `richText` fields within the collection.

As far as I understand, my only option now is to create a separate endpoint for this on the collection and then do the exclusion myself.

**Comment by w-tai** [comment]
I don't have an answer for this - but I'm looking for the exact same thing, not sure why it hasn't come up sooner. Other than a custom endpoint there doesn't seem to be a simple configuration available.

The use case is similar, only needing certain fields like title, description when returning a list in a collection, ignoring other data from the same level.

**Comment by w-tai** [comment]
It looks like the `afterRead` hook used inside a Collection is a good solution. The `findMany` boolean distinguishes between singular or list REST API query, and you can remove the fields manually from a copy of the `doc` object that you don't want to be returned in one or both default endpoints. You also have access to the `user` object, so you can bypass any manipulations when the Collection is queried via the Admin UI.

https://payloadcms.com/docs/hooks/collections#afterread

```typescript
// file: `src/collections/Something`
export const Something: CollectionConfig = {
  fields: [
    ...
  ]
  ...
  hooks: {
    afterRead: [
      (async ({doc, req, query, findMany}) => {
        if (req.user != null) return doc; // ignore hook for logged in users

        const docCopy = {...doc};

        // do not return createdAt field for any api calls
        // delete docCopy.createdAt;

        if (findMany) {
          // delete fields when listing collection
          delete docCopy.richText;
        } else {
          // delete fields when single item is queried
        }

        return docCopy;
      }) as CollectionAfterReadHook
    ],
  },
  ...
}
```

**Reply by revnelson** [reply]
This is great, but still requires the full document to be sent from the database to payload, a copy to be made, and the document to be trimmed. Imagine a collection of thousands of items, hundreds of fields each, on a site that gets tens of thousands of hits each day. A way to select the fields you need *before* the database request is made seems absolutely necessary.

**Comment by r1tsuu** [comment]
[This](https://github.com/teunmooij/payload-tools/blob/main/packages/query/README.md) plugin contains Select - Rest API

From readme:
After the plugin is enabled you can use a select query parameter in the same way you would with the where parameter:
```
http://localhost:3000/api/posts?select[color]=true
```

Is it? I didn't use but sadly probably works only for 0 depth (so can't exclude field inside blocks array relations etc), i'll check it
I think this should be included in payload core query functionality
